server/src/main/java/org/apache/cloudstack/network/lb/LoadBalancerConfigManagerImpl.java

```diff
@@ -0,0 +1,396 @@
+    }
+
+    private void checkPermission(Scope scope, Long networkId, Long vpcId, Long loadBalancerId, Boolean listAll) {
+        Account caller = CallContext.current().getCallingAccount();
+        if (scope == Scope.Network) {
+            if (networkId == null) {
+                if (listAll) {
+                    return;
+                }
+                throw new InvalidParameterValueException("networkId is required");
+            }
+            if (vpcId != null || loadBalancerId != null) {
+                throw new InvalidParameterValueException("vpcId and loadBalancerId should be null if scope is Network");
+            }
+            if (networkId == null) {
+                throw new InvalidParameterValueException("networkId is required");
+            }
+            NetworkVO network = _networkDao.findById(networkId);
+            if (network == null) {
+                throw new InvalidParameterValueException("Cannot find network by id " + networkId);
+            }
+            // Perform permission check
+            _accountMgr.checkAccess(caller, null, true, network);
+            if (network.getVpcId() != null) {
+                throw new InvalidParameterValueException("network " + network.getName() + " is a VPC tier, please add LB configs to VPC instead");
+            }
+        } else if (scope == Scope.Vpc) {
+            if (vpcId == null) {
+                if (listAll) {
+                    return;
+                }
+                throw new InvalidParameterValueException("vpcId is required");
+            }
+            if (networkId != null || loadBalancerId != null) {
+                throw new InvalidParameterValueException("networkId and loadBalancerId should be null if scope is Vpc");
+            }
+            VpcVO vpc = _vpcDao.findById(vpcId);
+            if (vpc == null) {
+                throw new InvalidParameterValueException("Cannot find vpc by id " + vpcId);
+            }
+            // Perform permission check
+            _accountMgr.checkAccess(caller, null, true, vpc);
+        } else if (scope == Scope.LoadBalancerRule) {
+            if (loadBalancerId == null) {
+                if (listAll) {
+                    return;
+                }
+                throw new InvalidParameterValueException("loadBalancerId is required");
+            }
+            if (networkId != null || vpcId != null) {
+                throw new InvalidParameterValueException("networkId and vpcId should be null if scope is LoadBalancerRule");
+            }
+            LoadBalancerVO rule = _lbDao.findById(loadBalancerId);
+            if (rule == null) {
+                throw new InvalidParameterValueException("Cannot find load balancer rule by id " + loadBalancerId);
+            }
+            if (networkId != null) {
+                // Perform permission check
+                checkPermission(Scope.Network, rule.getNetworkId(), null, null);
+            }
+        }
```

Comment:
This code seems repeated, but changing parameters; We could extract it to a method.